velox/exec/PrefixSort.cpp

```diff
@@ -129,28 +143,43 @@ compareByWord(uint64_t* left, uint64_t* right, int32_t bytes) {
 PrefixSortLayout PrefixSortLayout::makeSortLayout(
     const std::vector<TypePtr>& types,
     const std::vector<CompareFlags>& compareFlags,
-    uint32_t maxNormalizedKeySize) {
+    uint32_t maxNormalizedKeySize,
+    int32_t stringPrefixLength) {
   const uint32_t numKeys = types.size();
   std::vector<uint32_t> prefixOffsets;
   prefixOffsets.reserve(numKeys);
+  std::vector<uint32_t> encodeSizes;
+  encodeSizes.reserve(numKeys);
   std::vector<PrefixSortEncoder> encoders;
   encoders.reserve(numKeys);
 
   // Calculate encoders and prefix-offsets, and stop the loop if a key that
-  // cannot be normalized is encountered.
+  // cannot be normalized is encountered or only partial data of a key is
+  // normalized.
   uint32_t normalizedKeySize{0};
   uint32_t numNormalizedKeys{0};
+
+  bool lastKeyInPrefixIsPartial{false};
   for (auto i = 0; i < numKeys; ++i) {
     const std::optional<uint32_t> encodedSize =
-        PrefixSortEncoder::encodedSize(types[i]->kind());
+        PrefixSortEncoder::encodedSize(types[i]->kind(), stringPrefixLength);
     if (!encodedSize.has_value() ||
         normalizedKeySize + encodedSize.value() > maxNormalizedKeySize) {
       break;
     }
     prefixOffsets.push_back(normalizedKeySize);
     encoders.push_back({compareFlags[i].ascending, compareFlags[i].nullsFirst});
+    encodeSizes.push_back(encodedSize.value());
     normalizedKeySize += encodedSize.value();
     ++numNormalizedKeys;
```

Comment:
In function`comparePartNormalizedKeys`, https://github.com/facebookincubator/velox/blob/main/velox/exec/PrefixSort.cpp#L188, we compare the key after `numNormalizedKeys`, we need to compare the last StringType, so we should move `++numNormalizedKeys;` to L183, and test it in makeSortLayout, and rerun the benchmark.

Reply:
Updated benchmark result in PR description also.